### PR TITLE
feat(): FSRS 분석 & 테스트케이스 작성

### PR DESCRIPTION
**Pull request by Jaeja-Phil**
- fix & refactor SchedulingCard
- add missing updateState for fsrs repeat logic
- add test case 
 <div id='description'>
<h3>Summary by Bito</h3>
The changes focus on enhancing the FSRS implementation, including refactoring SchedulingCard for efficiency, adding comprehensive test suites for FSRS, Card, and SchedulingCard classes. The updates cover various scenarios such as card creation, state updates, scheduling, and log recording. Optimizations in ReviewLog creation and improvements in code structure are also included.
<br>
<br>
<b>Code change type</b>: Tests, Refactoring, Optimization
<br>
<br>
<b>Unit tests added</b>: True
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 3
</div>

### CODE REVIEW

**Comment by Jaeja-Phil** [comment]
<div>
<h2>Code Review Agent Run #8da850</h2>
<div>
<ul>

<li>
<b>AI Based Review: </b><small>✔️</small>&nbsp;Successful
</li>

</ul>
</div>
<div>
<h3>High-level Feedback</h3>
The code demonstrates a strong focus on testing and optimization, which is commendable. The enhanced rollback test and extended 'Easy' grade test show attention to detail and long-term behavior. To further improve, consider addressing the potential flaky test in CardTest by using a more precise comparison method for timestamps. Additionally, while the test coverage is extensive, ensure that all edge cases are covered, particularly in the FSRS algorithm's long-term behavior. Overall, the changes reflect a thoughtful approach to improving the FSRS implementation.
</div>
<br>

<div>
<h3>Actionable Issues</h3>

<div id="issue">
<div>
:page_facing_up: fsrs/src/test/kotlin/com/inout/fsrs/FSRSTest.kt
</div>
<details>
<summary>
Issues: Total - 2, High importance - 0
</summary>
<div>


<table width="100%">
<tbody>

<tr>
<td width="30%">
<a href ="https://github.com/in-out-team/api-server/pull/10/files#r1764868478">Line 236-236</a>
</td>
<td width="70%">
 <small>🟠</small> Medium importance - 1 &nbsp;&nbsp; 
</td>
</tr>

<tr>
<td width="30%">
<a href ="https://github.com/in-out-team/api-server/pull/10/files#r1764868495">Line 308-308</a>
</td>
<td width="70%">
 <small>🟠</small> Medium importance - 1 &nbsp;&nbsp; 
</td>
</tr>

</tbody>
</table>

</div>
</details>
</div>

<div id="issue">
<div>
:page_facing_up: fsrs/src/test/kotlin/com/inout/fsrs/model/CardTest.kt
</div>
<details>
<summary>
Issues: Total - 1, High importance - 0
</summary>
<div>


<table width="100%">
<tbody>

<tr>
<td width="30%">
<a href ="https://github.com/in-out-team/api-server/pull/10/files#r1764868509">Line 43-43</a>
</td>
<td width="70%">
 <small>🟠</small> Medium importance - 1 &nbsp;&nbsp; 
</td>
</tr>

</tbody>
</table>

</div>
</details>
</div>

</div>




<span><i>AI Code Review powered by</i> <sub><sub><a href="https://bito.ai/" target="_blank">Bito Logo</a></sub></sub></span>

</div>

File: fsrs/src/test/kotlin/com/inout/fsrs/FSRSTest.kt [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+            assertEquals(State.Review, easyLog.card.state)
+            assertEquals(secondDue.addDays(5), easyLog.card.due)
+            assertEquals(0, easyLog.card.elapsedDays)
+            assertEquals(5, easyLog.card.scheduledDays)
+            assertEquals(2, easyLog.card.reps)
+            assertEquals(0, easyLog.card.lapses)
+            assertEquals(secondDue, easyLog.card.lastReview)
+        }
+
+        @Test
```
Comment (Jaeja-Phil) [inline fsrs/src/test/kotlin/com/inout/fsrs/FSRSTest.kt:236]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/in-out-team/api-server/pull/10#issuecomment-2358174366>#8da850</a><span> - 09/18/2024, 11:15 am</span>
<br>

<div><small>🟠</small> Medium importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The test case for the 'Easy' grade doesn't adequately test the long-term behavior of the FSRS algorithm
</div>
<div id="fix">
<b>Fix:</b> Extend the test case to cover multiple repetitions and verify the long-term behavior of the FSRS algorithm for the 'Easy' grade
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -236,52 +236,77 @@
      @Test
      fun 'repeat with easy'() {
          // given
          val fsrs = createFSRS()
          val card = createCard()
          val firstDue =
              Calendar.getInstance().apply {
                  set(2024, 0, 1, 0, 0, 0)
              }.time
 
          // when
          val firstLogs = fsrs.repeat(card, firstDue).logs
          val secondDue = firstLogs[Grade.Easy]!!.card.due
          val secondLogs = fsrs.repeat(firstLogs[Grade.Easy]!!.card, secondDue).logs
 
          // then
          val againLog = secondLogs[Grade.Again]!!
          assertEquals(State.Relearning, againLog.card.state)
          assertEquals(secondDue.addMinutes(5), againLog.card.due)
          assertEquals(11, againLog.card.elapsedDays)
          assertEquals(0, againLog.card.scheduledDays)
          assertEquals(2, againLog.card.reps)
          assertEquals(1, againLog.card.lapses)
          assertEquals(secondDue, againLog.card.lastReview)
 
          val hardLog = secondLogs[Grade.Hard]!!
          assertEquals(State.Review, hardLog.card.state)
          assertEquals(secondDue.addDays(18), hardLog.card.due)
          assertEquals(11, hardLog.card.elapsedDays)
          assertEquals(18, hardLog.card.scheduledDays)
          assertEquals(2, hardLog.card.reps)
          assertEquals(0, hardLog.card.lapses)
          assertEquals(secondDue, hardLog.card.lastReview)
 
          val goodLog = secondLogs[Grade.Good]!!
          assertEquals(State.Review, goodLog.card.state)
          assertEquals(secondDue.addDays(42), goodLog.card.due)
          assertEquals(11, goodLog.card.elapsedDays)
          assertEquals(42, goodLog.card.scheduledDays)
          assertEquals(2, goodLog.card.reps)
          assertEquals(0, goodLog.card.lapses)
          assertEquals(secondDue, goodLog.card.lastReview)
 
          val easyLog = secondLogs[Grade.Easy]!!
          assertEquals(State.Review, easyLog.card.state)
          assertEquals(secondDue.addDays(98), easyLog.card.due)
          assertEquals(11, easyLog.card.elapsedDays)
          assertEquals(98, easyLog.card.scheduledDays)
          assertEquals(2, easyLog.card.reps)
          assertEquals(0, easyLog.card.lapses)
          assertEquals(secondDue, easyLog.card.lastReview)
 
 +        // Additional repetitions for long-term behavior
 +        var currentCard = easyLog.card
 +        var currentDue = currentCard.due
 +        for (i in 1..5) {
 +            val nextLogs = fsrs.repeat(currentCard, currentDue).logs
 +            val nextEasyLog = nextLogs[Grade.Easy]!!
 +            currentCard = nextEasyLog.card
 +            currentDue = currentCard.due
 +
 +            assertEquals(State.Review, currentCard.state)
 +            assertTrue(currentCard.scheduledDays > 98 * (i + 1))
 +            assertEquals(i + 2, currentCard.reps)
 +            assertEquals(0, currentCard.lapses)
 +            assertEquals(currentDue.addDays(-currentCard.scheduledDays), currentCard.lastReview)
 +        }
      }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [x] it was incorrectly flagged

File: fsrs/src/test/kotlin/com/inout/fsrs/FSRSTest.kt [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+                Calendar.getInstance().apply {
+                    set(2024, 0, 1, 0, 0, 0)
+                }.time
+            val firstRepeat = fsrs.repeat(card, due)
+
+            // when & then
+            listOf(Grade.Again, Grade.Hard, Grade.Good, Grade.Easy).forEach {
+                val recordLogItem = firstRepeat.logs[it]!!
+                val rollbackCard = fsrs.rollback(recordLogItem.card, recordLogItem.log)
+                assertEquals(cardCopied, rollbackCard)
```
Comment (Jaeja-Phil) [inline fsrs/src/test/kotlin/com/inout/fsrs/FSRSTest.kt:308]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/in-out-team/api-server/pull/10#issuecomment-2358174366>#8da850</a><span> - 09/18/2024, 11:15 am</span>
<br>

<div><small>🟠</small> Medium importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The rollback test case does not verify individual properties of the card after rollback
</div>
<div id="fix">
<b>Fix:</b> Enhance the rollback test to verify individual properties of the card after rollback
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -306,7 +306,15 @@
          listOf(Grade.Again, Grade.Hard, Grade.Good, Grade.Easy).forEach {
              val recordLogItem = firstRepeat.logs[it]!!
              val rollbackCard = fsrs.rollback(recordLogItem.card, recordLogItem.log)
 -            assertEquals(cardCopied, rollbackCard)
 +            assertEquals(cardCopied.state, rollbackCard.state, "State should match after rollback")
 +            assertEquals(cardCopied.due, rollbackCard.due, "Due date should match after rollback")
 +            assertEquals(cardCopied.stability, rollbackCard.stability, "Stability should match after rollback")
 +            assertEquals(cardCopied.difficulty, rollbackCard.difficulty, "Difficulty should match after rollback")
 +            assertEquals(cardCopied.elapsedDays, rollbackCard.elapsedDays, "Elapsed days should match after rollback")
 +            assertEquals(cardCopied.scheduledDays, rollbackCard.scheduledDays, "Scheduled days should match after rollback")
 +            assertEquals(cardCopied.reps, rollbackCard.reps, "Reps should match after rollback")
 +            assertEquals(cardCopied.lapses, rollbackCard.lapses, "Lapses should match after rollback")
 +            assertEquals(cardCopied.lastReview, rollbackCard.lastReview, "Last review should match after rollback")
          }
      }
  }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [x] it was incorrectly flagged

File: fsrs/src/test/kotlin/com/inout/fsrs/model/CardTest.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        fun `should create an empty card with current date if no date is given`() {
+            // Given
+            val now = Date()
+
+            // When
+            val card = Card.createEmptyCard()
+
+            // Then
+            assertEquals(State.New, card.state)
+            assertEquals(now.time / 1000, card.due.time / 1000)
```
Comment (Jaeja-Phil) [inline fsrs/src/test/kotlin/com/inout/fsrs/model/CardTest.kt:43]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/in-out-team/api-server/pull/10#issuecomment-2358174366>#8da850</a><span> - 09/18/2024, 11:15 am</span>
<br>

<div><small>🟠</small> Medium importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The comparison of timestamps using integer division by 1000 may lead to flaky tests due to potential small time differences.
</div>
<div id="fix">
<b>Fix:</b> Use a more precise comparison method for timestamps or mock the time in the test to ensure consistent results.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -35,9 +35,11 @@
      fun 'should create an empty card with current date if no date is given'() {
          // Given
          val now = Date()
 +        val tolerance = 100 // 100 milliseconds tolerance

          // When
          val card = Card.createEmptyCard()

          // Then
          assertEquals(State.New, card.state)
 -        assertEquals(now.time / 1000, card.due.time / 1000)
 +        assertTrue(Math.abs(now.time - card.due.time) < tolerance)
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged
